### ISSUE TITLE
Issues in configuring AzureCP, i.e. adding a tenant

### ISSUE THREAD

**Issue by powerfischi** [Yvand/AzureCP#135]
Environment: SharePoint 2019 (Build Version 16.0.10372.20060), Windows Server 2019

When trying to add a new tenant via /_admin/azurecp/AzureCPSettings.aspx and hitting the button "Test connection to tenant", connection to Central Admin is resetted, resulting in an empty page. The Windows Event Log states:

> A process serving application pool 'SharePoint Central Administration v4' suffered a fatal communication error with the Windows Process Activation Service. The process id was '15560'. The data field contains the error number.

So I tried adding the tenant via PowerShell. Then Central Admin is showing the configured tenant. However, this makes the Security Token Service fault during authentication via the configured SPTrustedIdentityTokenIssuer.

I checked ULS and found this related exception:

>Claims Saml Sign-In: Could not get local token for trusted third party token. Exception: 'System.ServiceModel.CommunicationException: An error occurred while receiving the HTTP response to http://localhost:32843/SecurityTokenServiceApplication/securitytoken.svc. This could be due to the service endpoint binding not using the HTTP protocol. This could also be due to an HTTP request context being aborted by the server (possibly due to the service shutting down). See server logs for more details. ---> System.Net.WebException: The underlying connection was closed: An unexpected error occurred on a receive. ---> System.IO.IOException: Unable to read data from the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host    
> at System.Net.Sockets.Socket.Receive(Byte[] buffer, Int32 offset, Int32 size, SocketFlags socketFlags)    
> at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)     -
> -- End of inner exception stack trace ---    
> at System.Net.Sockets.NetworkStream.Read(Byte[] buffer, Int32 offset, Int32 size)    
> at System.Net.PooledStream.Read(Byte[] buffer, Int32 offset, Int32 size)    
> at System.Net.Connection.SyncRead(HttpWebRequest request, Boolean userRetrievedStream, Boolean probeRead)     -
> -- End of inner exception stack trace ---    
> at System.Net.HttpWebRequest.GetResponse()    
> at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)     -
> -- End of inner exception stack trace ---    Server stack trace:     
> at System.ServiceModel.Channels.HttpChannelUtilities.ProcessGetResponseWebException(WebException webException, HttpWebRequest request, HttpAbortReason abortReason)    
> at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)    
> at System.ServiceModel.Channels.RequestChannel.Request(Message message, TimeSpan timeout)    
> at System.ServiceModel.Dispatcher.RequestChannelBinder.Request(Message message, TimeSpan timeout)    
> at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)    
> at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs)    
> at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(IMethodCallMessage methodCall, ProxyOperationRuntime operation)    
> at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(IMessage message)    Exception rethrown
> at [0]:     
> at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)    
> at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)    
> at Microsoft.IdentityModel.Protocols.WSTrust.IWSTrustContract.Issue(Message message)    
> at Microsoft.IdentityModel.Protocols.WSTrust.WSTrustChannel.Issue(RequestSecurityToken rst, RequestSecurityTokenResponse& rstr)    
> at Microsoft.IdentityModel.Protocols.WSTrust.WSTrustChannel.Issue(RequestSecurityToken rst)    
> at Microsoft.SharePoint.SPSecurityContext.SecurityTokenForContext(Uri context, Boolean bearerToken, SecurityToken onBehalfOf, SecurityToken actAs, SecurityToken delegateTo, SPRequestSecurityTokenProperties properties)    
> at Microsoft.SharePoint.SPSecurityContext.SecurityTokenForOnBehalfOfContext(Uri context, SecurityToken onBehalfOf)    
> at Microsoft.SharePoint.IdentityModel.SPFederationAuthenticationModule.ExchangeArgumentTrustedThirdPartySessionSecurityTokenForLocalToken(SecurityToken thirdPartyToken, SessionSecurityTokenCreatedEventArgs arguments)'. Stack: '  Server stack trace:     
> at System.ServiceModel.Channels.HttpChannelUtilities.ProcessGetResponseWebException(WebException webException, HttpWebRequest request, HttpAbortReason abortReason)    
> at System.ServiceModel.Channels.HttpChannelFactory`1.HttpRequestChannel.HttpChannelRequest.WaitForReply(TimeSpan timeout)    
> at System.ServiceModel.Channels.RequestChannel.Request(Message message, TimeSpan timeout)    
> at System.ServiceModel.Dispatcher.RequestChannelBinder.Request(Message message, TimeSpan timeout)    
> at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs, TimeSpan timeout)    
> at System.ServiceModel.Channels.ServiceChannel.Call(String action, Boolean oneway, ProxyOperationRuntime operation, Object[] ins, Object[] outs)    
> at System.ServiceModel.Channels.ServiceChannelProxy.InvokeService(IMethodCallMessage methodCall, ProxyOperationRuntime operation)    
> at System.ServiceModel.Channels.ServiceChannelProxy.Invoke(IMessage message)    Exception rethrown
> at [0]:     
> at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)    
> at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)    
> at Microsoft.IdentityModel.Protocols.WSTrust.IWSTrustContract.Issue(Message message)    
> at Microsoft.IdentityModel.Protocols.WSTrust.WSTrustChannel.Issue(RequestSecurityToken rst, RequestSecurityTokenResponse& rstr)    
> at Microsoft.IdentityModel.Protocols.WSTrust.WSTrustChannel.Issue(RequestSecurityToken rst)    
> at Microsoft.SharePoint.SPSecurityContext.SecurityTokenForContext(Uri context, Boolean bearerToken, SecurityToken onBehalfOf, SecurityToken actAs, SecurityToken delegateTo, SPRequestSecurityTokenProperties properties)    
> at Microsoft.SharePoint.SPSecurityContext.SecurityTokenForOnBehalfOfContext(Uri context, SecurityToken onBehalfOf)    
> at Microsoft.SharePoint.IdentityModel.SPFederationAuthenticationModule.ExchangeArgumentTrustedThirdPartySessionSecurityTokenForLocalToken(SecurityToken thirdPartyToken, SessionSecurityTokenCreatedEventArgs arguments)'.

AzureCP itself is not logging much.

> 11.15.2021 22:42:45.21	w3wp.exe (0x23B8)	0x3928	AzureCP	Core	1337	Medium	[AzureCP] Configuration 'AzureCPConfig' changed to version 2842103, refreshing local copy	a37b03a0-97ed-1078-8d86-83d590a2ebba
> 11.15.2021 23:00:35.21	w3wp.exe (0x29C0)	0x23B4	AzureCP	Core	1337	Medium	[AzureCP] Configuration 'AzureCPConfig' changed to version 2842103, refreshing local copy	a87c03a0-f7fa-1078-8d86-884e27b82e01
> 11.15.2021 23:20:32.92	w3wp.exe (0x2D04)	0x1CFC	AzureCP	Core	1337	Medium	[AzureCP] Configuration 'AzureCPConfig' changed to version 2842103, refreshing local copy	cd7d03a0-77eb-1078-8d86-8c459a2b4812

A soon as I remove the tenant from AzureCP configuration, authentication is working again. No faulty Security Token Service.

Connectivity to Azure was tested as outlined in the docs.
https://azurecp.yvand.net/docs/help/troubleshooting/

And now I'm kinda stuck and appreciate any hints.

Thanks in advance.

**Comment by Yvand** [comment]
@powerfischi based on the the behavior, I think AzureCP tries to get the group membership from Azure AD when users  sign-in, but somehow it does not work and it times out.
You can investigate the cause by looking at the SharePoint logs and/or by testing the connection to Azure AD from a SharePoint server. Both are explained in [this page](https://azurecp.yvand.net/docs/help/troubleshooting/).

**Comment by powerfischi** [comment]
Thanks. I checked general connectivity. All good. Now I need to find the time to analyze the request with fiddler.

**Comment by powerfischi** [comment]
Now I found the time to do the troubleshooting with fiddler. It is somehow strange. I reconfigured web.config for Central Admin and my Content Web Application to use fiddler as a proxy, as mentioned in the troubleshooting section. But fiddler is capturing nothing from AzureCP. I don't see any outgoing connection towards MS Graph.

**Comment by Yvand** [comment]
@powerfischi I just updated [the documentation](https://azurecp.yvand.net/docs/help/troubleshooting/#inspect-the-traffic-to-azure-ad) to mention explicitly that you need to set the filter in Fiddler to "All Processes" or "Non-Browsers". Did you do that?

**Comment by powerfischi** [comment]
Yes, I did this. So I tried resetting the AzureCP configuration and starting from scratch. When adding a tenant from Central Admin and pushing the "Test connection to tenant" button the application pool crashes.

I got this in the event log:
> A process serving application pool 'SharePoint Central Administration v4' suffered a fatal communication error with the Windows Process Activation Service. The process id was '14036'. The data field contains the error number.

ULS has no entries for AzureCP:
```
Merge-SPLogFile -Path "AzureCP_logging.log" -Overwrite -Area "AzureCP" -StartTime (Get-Date).AddMinutes(-10)
WARNING: Cmdlet did not return any records in the log file. Check your time range or filters.
```

```
Get-SPLogLevel| ?{$_.Area -like "AzureCP"}

Area                            Name                            TraceSev   EventSev
----                            ----                            --------   --------
AzureCP                         Augmentation                    Verbose    Error
AzureCP                         Claims Picking                  Verbose    Error
AzureCP                         Configuration                   Verbose    Error
AzureCP                         Core                            Verbose    Error
AzureCP                         Custom                          Verbose    Error
AzureCP                         Debug                           Verbose    Error
AzureCP                         Lookup                          Verbose    Error
AzureCP                         Rehydration                     Verbose    Error
```

**Comment by powerfischi** [comment]
Additional remark. At the same time I get this in the event log:

> Log Name:      Application
> Source:        Application Error
> Date:          23.11.2021 23:28:34
> Event ID:      1000
> Task Category: (100)
> Level:         Error
> Keywords:      Classic
> User:          N/A
> Computer:      xxxxxxxxxxxxx
> Description:
> Faulting application name: w3wp.exe, version: 10.0.17763.1, time stamp: 0xcfdb13d8
> Faulting module name: PerfMon64.dll, version: 8.0.11037.0, time stamp: 0x5d5fbab4
> Exception code: 0xc0000409
> Fault offset: 0x000000000013f68f
> Faulting process id: 0x4f8
> Faulting application start time: 0x01d7e0b9522b2413
> Faulting application path: c:\windows\system32\inetsrv\w3wp.exe
> Faulting module path: C:\Program Files\Microsoft Monitoring Agent\Agent\APMDOTNETAgent\V8.0.10918.0\PerfMon64.dll
> Report Id: 0fc4d9bd-a99a-4efc-b15e-8c00a953b418
> Faulting package full name: 
> Faulting package-relative application ID: 
> Faulting package-relative application ID:

**Comment by Yvand** [comment]
@powerfischi that's weird, then can you get the SharePoint logs filtered on the correlation ID of the POST request when you hit "Test connection to tenant" button ?

**Comment by powerfischi** [comment]
@Yvand the issue is solved. I found the culprit. :)

The farm member machines all had the SCOM agent installed. By default the agent is installed with APM enabled. However APM is not supported with SharePoint as described [here](http://sp2013.pro/2017/03/loading-assembly-produce-different-grant-set-instances/). Once I uninstalled the SCOM agent, AzureCP started working without issues.

Thanks for your support and patience. ;)
